fractal_server/app/runner/shutdown.py

```diff
@@ -0,0 +1,161 @@
+import time
+from logging import Logger
+
+from sqlmodel import select
+
+from fractal_server.app.models.v1 import ApplyWorkflow
+from fractal_server.app.models.v2 import JobV2
+from fractal_server.app.routes.aux._job import _write_shutdown_file
+from fractal_server.app.security import get_async_session_context
+from fractal_server.config import get_settings
+from fractal_server.syringe import Inject
+
+
+# If backend is not SLURM, skip this part
+# Note that this would need to take place for both V1 and V2 jobs.
+
+
+async def cleanup_after_shutdown(
+    jobsV1: list[int], jobsV2: list[int], logger: Logger
+):
+    logger.info("Cleanup function after shutdown")
+    try:
+        async with get_async_session_context() as session:
+            jobsV2_db = (
+                (
+                    await session.execute(
+                        select(JobV2)
+                        .where(JobV2.id.in_(jobsV2))
+                        .where(JobV2.status == "submitted")
+                    )
+                )
+                .scalars()
+                .all()
+            )  # untested
+
+            jobsV1_db = (
+                (
+                    await session.execute(
+                        select(ApplyWorkflow)
+                        .where(ApplyWorkflow.id.in_(jobsV1))
+                        .where(ApplyWorkflow.status == "submitted")
+                    )
+                )
+                .scalars()
+                .all()
+            )
+            for job in jobsV2_db:
+                _write_shutdown_file(job=job)
+
+            for job in jobsV1_db:
+                _write_shutdown_file(job=job)
+
+            settings = Inject(get_settings)
+
+            t_start = time.perf_counter()
+            while (
+                time.perf_counter() - t_start
+            ) < settings.FRACTAL_GRACEFUL_SHUTDOWN_TIME:  # 30 seconds
+                logger.info("Waiting 3 seconds before checking")
+                time.sleep(3)
+                jobsV2_db = (
+                    (
+                        await session.execute(
+                            select(JobV2)
+                            .where(JobV2.id.in_(jobsV2))
+                            .where(JobV2.status == "submitted")
+                        )
+                    )
+                    .scalars()
+                    .all()
+                )  # untested
+
+                jobsV1_db = (
+                    (
+                        await session.execute(
+                            select(ApplyWorkflow)
+                            .where(ApplyWorkflow.id.in_(jobsV1))
+                            .where(ApplyWorkflow.status == "submitted")
+                        )
+                    )
+                    .scalars()
+                    .all()
+                )
+
+                for job in jobsV2_db:
```

Comment:
I think these additional calls to `_write_shutdown_file` are not needed. The file is written once, at the beginning, and then we are just waiting for when the executor identifies it and triggers the rest of its own shutdown (so that the background task also marks jobs as failed in the database).